nomt/src/beatree/allocator/mod.rs

```diff
@@ -167,3 +172,242 @@ pub struct AllocatorCommitOutput {
+///
+/// The store is shadow-paged and makes use of an embedded free-list to track free pages.
+#[derive(Clone)]
+pub struct Store {
+    file: Arc<File>,
+    sync: Arc<Mutex<StoreSync>>,
+}
+
+impl Store {
+    /// Create a new `Store` over an existing file.
+    pub fn open(
+        page_pool: &PagePool,
+        file: File,
+        bump: PageNumber,
+        free_list_head: Option<PageNumber>,
+    ) -> anyhow::Result<Self> {
+        let file_size = file.metadata()?.size() as usize;
+
+        let sync = StoreSync {
+            free_list: FreeList::read(page_pool, &file, free_list_head)?,
+            bump,
+            max_bump: PageNumber((file_size / PAGE_SIZE) as u32),
+        };
+
+        Ok(Store {
+            file: Arc::new(file),
+            sync: Arc::new(Mutex::new(sync)),
+        })
+    }
+
+    /// Reads the page with the specified page number. Blocks the current thread.
+    pub fn query(&self, page_pool: &PagePool, pn: PageNumber) -> FatPage {
+        io::read_page(page_pool, &self.file, pn.0 as u64).unwrap()
+    }
+
+    /// Create an I/O command for querying a page by number.
+    pub fn io_command(&self, page_pool: &PagePool, pn: PageNumber, user_data: u64) -> IoCommand {
+        let page = page_pool.alloc_fat_page();
+        IoCommand {
+            kind: IoKind::Read(self.file.as_raw_fd(), pn.0 as u64, page),
+            user_data,
+        }
+    }
+
+    /// Start synchronization. This produces two handles,
+    /// a [`SyncAllocator`] and a [`SyncFinisher`].
+    ///
+    /// The `SyncAllocator` may be cloned and shared between threads to allocate pages. The finisher
+    /// should be used to update the embedded free-list and prepare the writes for that purpose.
+    ///
+    /// This will block if another sync is in progress.
+    pub fn start_sync(&self) -> (SyncAllocator, SyncFinisher) {
+        let sync = Mutex::lock_arc(&self.sync);
+        let (sync_tx, sync_rx) = crossbeam_channel::bounded(1);
+
+        let finisher = SyncFinisher {
+            file: self.file.clone(),
+            sync_finish: sync_rx,
+        };
+
+        let allocator = SyncAllocator {
+            file: self.file.clone(),
+            inner: Arc::new(SyncAllocatorInner {
+                max_bump: AtomicU32::new(sync.max_bump.0),
+                set_len_guard: Mutex::new(sync.max_bump),
+                sync: Some(sync),
+                send_to_finish: sync_tx,
+                allocations: AtomicUsize::new(0),
+            }),
+        };
+
+        (allocator, finisher)
+    }
+}
+
+struct StoreSync {
+    /// the next page number of the store.
+    bump: PageNumber,
+    /// the maximum capacity of the store. beyond this size, the store must have its
+    /// length extended.
+    max_bump: PageNumber,
+    /// the free-list of pages.
+    free_list: FreeList,
+}
+
+type StoreSyncGuard = ArcMutexGuard<parking_lot::RawMutex, StoreSync>;
+
+// Grow the store by 1MB at a time.
+const GROW_STORE_BY_PAGES: u32 = 256;
```

Comment:
This feels a bit low tbh. 

I don't think we should be a scrooge about disk space and we can afford to bump megabytes per call. 8-64 feels more appropriate.

Reply:
I agree! But this was the previous value and I didn't want to mix concerns. We should follow up by increasing this constant, especially because it occasionally interrupts the happy path.